syft/pkg/cataloger/cpe.go

```diff
@@ -546,68 +513,88 @@ func scanByHyphenOrUnderscore(data []byte, atEOF bool) (advance int, token []byt
 	return 0, nil, nil
 }
 
-func addSeparatorVariations(fields *strset.Set) {
-	for _, field := range fields.List() {
+func addDelimiterVariations(fields *cpeFieldCandidateSet) {
+	for _, candidate := range fields.list(filterCandidatesByDelimiterVariations) {
+		field := candidate.value
 		hasHyphen := strings.Contains(field, "-")
 		hasUnderscore := strings.Contains(field, "_")
 
 		if hasHyphen {
 			// provide variations of hyphen candidates with an underscore
-			fields.Add(strings.ReplaceAll(field, "-", "_"))
+			newValue := strings.ReplaceAll(field, "-", "_")
+			candidate.value = newValue
+			fields.add(candidate)
 		}
 
 		if hasUnderscore {
 			// provide variations of underscore candidates with a hyphen
-			fields.Add(strings.ReplaceAll(field, "_", "-"))
+			newValue := strings.ReplaceAll(field, "_", "-")
+			candidate.value = newValue
+			fields.add(candidate)
 		}
 	}
 }
 
-func candidateVendorsForRPM(p pkg.Package) (candidates []string) {
+func candidateVendorsForRPM(p pkg.Package) *cpeFieldCandidateSet {
 	metadata, ok := p.Metadata.(pkg.RpmdbMetadata)
 	if !ok {
 		return nil
 	}
 
+	vendors := newCPRFieldCandidateSet()
+
 	if metadata.Vendor != "" {
-		candidates = append(candidates, normalizeTitle(metadata.Vendor))
+		vendors.add(cpeFieldCandidate{
+			value:                 normalizeTitle(metadata.Vendor),
+			disallowSubSelections: true,
+		})
 	}
 
-	return candidates
+	return vendors
 }
 
-func candidateVendorsForPython(p pkg.Package) []string {
+func candidateVendorsForPython(p pkg.Package) *cpeFieldCandidateSet {
 	metadata, ok := p.Metadata.(pkg.PythonPackageMetadata)
 	if !ok {
 		return nil
 	}
 
-	candidates := strset.New()
+	vendors := newCPRFieldCandidateSet()
 
 	if metadata.Author != "" {
-		candidates.Add(normalizeName(metadata.Author))
+		vendors.add(cpeFieldCandidate{
```

Comment:
this is ripe to be changed to a option function pattern (functions named by the configuration options they control). This is a large change, but may be worth while.

Reply:
changed in 79c07df73c6ef8004dae5d821f7a1ce763ed7c76